cocos/audio/android/PcmData.cpp

```diff
@@ -0,0 +1,128 @@
+
+http://www.cocos2d-x.org
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+****************************************************************************/
+
+#define LOG_TAG "PcmData"
+
+#include "audio/android/OpenSLHelper.h"
+
+#include "audio/android/PcmData.h"
+
+PcmData::PcmData()
+{
+//    LOGD("In the constructor of PcmData (%p)", this);
+    reset();
+}
+
+PcmData::~PcmData()
+{
+//    LOGD("In the destructor of PcmData (%p)", this);
+    reset();
+}
+
+PcmData::PcmData(const PcmData& o)
+{
+//    LOGD("In the copy constructor of PcmData (%p)", this);
+    numChannels = o.numChannels;
+    sampleRate = o.sampleRate;
+    bitsPerSample = o.bitsPerSample;
+    containerSize = o.containerSize;
+    channelMask = o.channelMask;
+    endianness = o.endianness;
+    numFrames = o.numFrames;
+    pcmBuffer = o.pcmBuffer;
+}
+
+PcmData::PcmData(PcmData&& o)
+{
+//    LOGD("In the move constructor of PcmData (%p)", this);
+    numChannels = o.numChannels;
+    sampleRate = o.sampleRate;
+    bitsPerSample = o.bitsPerSample;
+    containerSize = o.containerSize;
+    channelMask = o.channelMask;
+    endianness = o.endianness;
+    numFrames = o.numFrames;
+    pcmBuffer = o.pcmBuffer;
+    o.reset();
+}
+
+PcmData& PcmData::operator= (const PcmData& o)
+{
+//    LOGD("In the copy assignment of PcmData");
+    numChannels = o.numChannels;
+    sampleRate = o.sampleRate;
+    bitsPerSample = o.bitsPerSample;
+    containerSize = o.containerSize;
+    channelMask = o.channelMask;
+    endianness = o.endianness;
+    numFrames = o.numFrames;
+    pcmBuffer = o.pcmBuffer;
+    return *this;
+}
+
+PcmData& PcmData::operator= (PcmData&& o)
+{
+//    LOGD("In the move assignment of PcmData");
+    numChannels = o.numChannels;
+    sampleRate = o.sampleRate;
+    bitsPerSample = o.bitsPerSample;
+    containerSize = o.containerSize;
+    channelMask = o.channelMask;
+    endianness = o.endianness;
+    numFrames = o.numFrames;
+    pcmBuffer = o.pcmBuffer;
```

Comment:
I would use `pcmBuffer = std::move( o.pcmBuffer )` here to prevent useless updates of the pointer's ref counter.

Also, without this move an unfortunate code like the following would result in an empty data:

    PcmData data;
    // […] initialize with valid data
    assert( data.isValid() );
    data = std::move( data )
    assert( data.isValid() ); // fails, data has been reset.